File: vane-core/src/main/java/org/oddlama/vane/core/config/recipes/RecipeList.java
package org.oddlama.vane.core.config.recipes;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.oddlama.vane.core.config.ConfigDictSerializable;

public class RecipeList implements ConfigDictSerializable {

    private List<RecipeDefinition> recipes = new ArrayList<>();

    public RecipeList() {}

    public RecipeList(List<RecipeDefinition> recipes) {
        this.recipes = recipes;
    }

    public List<RecipeDefinition> recipes() {
        return recipes;
    }

    // Mapear nombres especiales a PascalCase para YAML: 'generic' -> 'Generic', 'terralith_generic' -> 'TerralithGeneric', 'terralith_rare' -> 'TerralithRare', 'ancientcity' -> 'AncientCity', 'bastion' -> 'Bastion', 'from_shulker_box' -> 'FromShulkerBox'
    public Map<String, Object> toDict() {
        return recipes
            .stream()
            .collect(Collectors.toMap(r -> toYamlName(r.name()), RecipeDefinition::toDict));
    }

    public void fromDict(final Map<String, Object> dict) {
        recipes.clear();
        for (final var e : dict.entrySet()) {
            final var key = e.getKey();
            final String name = key == null ? null : key;
            final var internalName = name == null || name.isEmpty() ? name : fromYamlName(name);
            recipes.add(RecipeDefinition.fromDict(internalName, e.getValue()));
        }
    }

    public static RecipeList of(RecipeDefinition... defs) {
        final var rl = new RecipeList();
        rl.recipes = Arrays.asList(defs);
        return rl;
    }

    private static String toYamlName(final String s) {
        if (s == null || s.isEmpty()) return s;
        final var low = s.toLowerCase();
        switch (low) {
            case "generic":
                return "Generic";
            case "terralith_generic":
                return "TerralithGeneric";
            case "terralith_rare":
                return "TerralithRare";
            case "ancientcity":
                return "AncientCity";
            case "bastion":
                return "Bastion";
            case "from_shulker_box":
                return "FromShulkerBox";
            default:
                return s;
        }
    }

    private static String fromYamlName(final String s) {
        if (s == null || s.isEmpty()) return s;
        final var low = s.toLowerCase().replaceAll("[_\\s]", "");
        switch (low) {
            case "generic":
                return "generic";
            case "terralithgeneric":
                return "terralith_generic";
            case "terralithrare":
                return "terralith_rare";
            case "ancientcity":
                return "ancientcity";
            case "bastion":
                return "bastion";
            case "fromshulkerbox":
                return "from_shulker_box";
            default:
                return s;
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/config/recipes/RecipeList.kt
package org.oddlama.vane.core.config.recipes

import org.oddlama.vane.core.config.ConfigDictSerializable
import java.util.*

/**
 * Config-serializable list wrapper for recipe definitions.
 *
 * @param recipes initial recipe definitions.
 */
class RecipeList(private var recipes: MutableList<RecipeDefinition?> = mutableListOf()) : ConfigDictSerializable {

    /** Returns the mutable list of recipe definitions. */
    fun recipes(): MutableList<RecipeDefinition?> = recipes

    /** Serializes this recipe list to dictionary form. */
    override fun toDict(): MutableMap<String, Any> =
        recipes.filterNotNull().associateTo(LinkedHashMap()) { r -> toYamlName(r.name)!! to (r.toDict() as Any) }

    /** Loads recipe definitions from dictionary form. */
    override fun fromDict(dict: MutableMap<String, Any>) {
        recipes.clear()
        dict.entries.forEach { (key, value) ->
            val internalName = if (key.isEmpty()) key else fromYamlName(key)
            recipes.add(RecipeDefinition.fromDict(internalName, value))
        }
    }

    /**
     * Construction and name-mapping helpers.
     */
    companion object {
        @JvmStatic
                /** Creates a [RecipeList] from vararg definitions. */
        fun of(vararg defs: RecipeDefinition?): RecipeList = RecipeList(defs.toMutableList())

        /** Maps internal recipe names to preferred YAML display names. */
        private fun toYamlName(s: String?): String? {
            if (s.isNullOrEmpty()) return s
            val low = s.lowercase(Locale.getDefault())
            return when (low) {
                "generic" -> "Generic"
                "terralith_generic" -> "TerralithGeneric"
                "terralith_rare" -> "TerralithRare"
                "ancientcity" -> "AncientCity"
                "bastion" -> "Bastion"
                "from_shulker_box" -> "FromShulkerBox"
                else -> s
            }
        }

        /** Maps YAML display names back to internal recipe names. */
        private fun fromYamlName(s: String?): String? {
            if (s.isNullOrEmpty()) return s
            val low = s.lowercase(Locale.getDefault()).replace("[_\\s]".toRegex(), "")
            return when (low) {
                "generic" -> "generic"
                "terralithgeneric" -> "terralith_generic"
                "terralithrare" -> "terralith_rare"
                "ancientcity" -> "ancientcity"
                "bastion" -> "bastion"
                "fromshulkerbox" -> "from_shulker_box"
                else -> s
            }
        }
    }
}
